src/Azure/Orleans.GrainDirectory.AzureStorage/AzureTableGrainDirectory.cs

```diff
@@ -118,24 +123,27 @@ public Task UnregisterSilos(List<SiloAddress> siloAddresses)
 
         private async Task UnregisterManyBlock(List<GrainAddress> addresses)
         {
-            var pkFilter = TableQuery.GenerateFilterCondition(nameof(GrainDirectoryEntity.PartitionKey), QueryComparisons.Equal, this.clusterId);
-            string rkFilter = TableQuery.CombineFilters(
-                    TableQuery.GenerateFilterCondition(nameof(GrainDirectoryEntity.RowKey), QueryComparisons.Equal, GrainDirectoryEntity.GrainIdToRowKey(addresses[0].GrainId)),
-                    TableOperators.And,
-                    TableQuery.GenerateFilterCondition(nameof(GrainDirectoryEntity.ActivationId), QueryComparisons.Equal, addresses[0].ActivationId)
-                    );
-
-            foreach (var addr in addresses.Skip(1))
+            var queryBuilder = new StringBuilder();
+            queryBuilder.Append(TableClient.CreateQueryFilter($"(PartitionKey eq {clusterId}) and ("));
+            var first = true;
+            foreach (var addr in addresses)
             {
-                var tmp = TableQuery.CombineFilters(
-                    TableQuery.GenerateFilterCondition(nameof(GrainDirectoryEntity.RowKey), QueryComparisons.Equal, GrainDirectoryEntity.GrainIdToRowKey(addr.GrainId)),
-                    TableOperators.And,
-                    TableQuery.GenerateFilterCondition(nameof(GrainDirectoryEntity.ActivationId), QueryComparisons.Equal, addr.ActivationId)
-                    );
-                rkFilter = TableQuery.CombineFilters(rkFilter, TableOperators.Or, tmp);
+                if (!first)
+                {
+                    queryBuilder.Append(" or ");
+                }
+                else
+                {
+                    first = false;
+                }
+
+                var rowKey = GrainDirectoryEntity.GrainIdToRowKey(addr.GrainId);
+                var queryClause = TableClient.CreateQueryFilter($"((RowKey eq {rowKey}) and (ActivationId eq {addr.ActivationId}))");
```

Comment:
nit: If you are including the PartitionKey and RowKey, no additional properties are needed. PK and RK together are globally unique given the same table.

Also - depending on how large this partition can get, you may want to see if you'd get better performance from multiple (perhaps in parallel) single resource requests (`GetEntitiyAsync`).

Reply:
> Also - depending on how large this partition can get, you may want to see if you'd get better performance from multiple (perhaps in parallel) single resource requests (GetEntitiyAsync).

What's the deciding factor between the two? Should we be batching into blocks of 100 or 200 point deletions and issuing it as a transaction instead?

Reply:
> nit: If you are including the PartitionKey and RowKey, no additional properties are needed. PK and RK together are globally unique given the same table.

The reason for the additional filter is that ActivationId may have been updated to something else. It's similar to an ETag. We only want to delete if the PK, RK, and ActivationId match.

Reply:
I don't know of any official guidance. I'd do some simple perf tests with realistic partition sizes to determine if it's worth while. I'm not even sure if two point queries is faster than 1 with a (PK eq 'a' and RK eq 'b' or PK eq 'c' and RK eq 'd'). I just know that the most efficiently indexed request possible is GetEntity, which does a different request than a query ([this](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/tables/Azure.Data.Tables/tests/SessionRecords/TableClientLiveTests(Storage)/GetEntityReturnsSingleEntity.json#L47) vs [this](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/tables/Azure.Data.Tables/tests/SessionRecords/TableClientLiveTests(Storage)/QueryReturnsEntitiesWithoutOdataAnnoations.json#L101))